### PR TITLE
Fix composer community and channel selection

### PR DESCRIPTION
**Pull request by brianlovin**
**Status**
- [ ] WIP
- [x] Ready for review
- [ ] Needs testing

**Deploy after merge (delete what needn't be deployed)**
- hyperion (frontend)

Oh boy look at this unit of a test file. @mxstbr if e2e tests are failing here it's *highly* likely that it's because of rate limiting (see #4750)

Otherwise - give this a pull, poke around and see if you can break it by putting in invalid community/channel ids, ids of communities/channels where you're not a member, etc.

### CODE REVIEW

**Comment by brianlovin** [comment]
As far as I can tell, the `test_web` failures are due to rate limiting and flaky tests. Everything passes locally - famous last words :P

File: cypress/integration/composer_selection_spec.js [hunk 1]
```diff
@@ -0,0 +1,369 @@
...
+    channelDropdownIsHidden();
+
+    communityDropdown().select('Spectrum');
+    channelDropdownIsEnabled();
+    channelDropdown().contains('General');
+    channelDropdown().contains('Private');
+
+    communityDropdown().select('Payments');
+    channelDropdown().contains('General');
+    channelDropdown().contains('Private');
```
Comment (mxstbr) [inline cypress/integration/composer_selection_spec.js:116]: We should add a third channel to Payments to make sure it fetched the new channels

File: src/components/composer/LocationSelectors/CommunitySelector.js [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+  if (loading) return <LoadingSelect />;
+
+  /*
+    If for some reason the query to get an individual community fails,
+    just return the AvailableCommunitiesDropdown which will do a followup
+    fetch of all the current user's communities
+  */
+  if (error) {
+    // remove any parent selected community id
+    onCommunityChange('');
```
Comment (mxstbr) [inline src/components/composer/LocationSelectors/CommunitySelector.js]: We should move this to a `React.useEffect`, something like

```JS
React.useEffect(() => {
  if (error) onCommunityChange('')
}, [error]);

if (error) return <Available...
```

File: src/components/composer/LocationSelectors/index.js [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+  /*
+    Whenever the browser location.search changes, check for query parameters 
+    related to the composer and update the state of the composer.
+  */
+  useEffect(
+    () => {
+      setStateFromQueryParams();
+    },
+    [location.search]
+  );
```
Comment (mxstbr) [inline src/components/composer/LocationSelectors/index.js:50]: Super great use of `useEffect`! 😍 

I think we do not need to one above, I am fairly certain all useEffects run once on mount and then whenever their dependencies (second arg) change. So I think we can remove the one above with `[]` as deps and just have this one!

**Review by mxstbr — COMMENTED** [review]
This is such an upgrade, wow! 😍😍😍😍😍 Great job on the treatment with the pre-selected communities and channels, that makes it really clear that it's locked in-place. 

I really really like this, amazing work. I have a couple of code nitpicks inline, but overall this is such a massive upgrade—I cannot wait to ship this!

**Comment by brianlovin** [comment]
Fixed all the nits, good call :) Once tests pass I’ll merge and get this up.

**Comment by brianlovin** [comment]
This one failing e2e test is driving me nuts. It works perfectly locally, run as an isolated test and run within the full suite. I have no clue why it's failing in CI only...

**Comment by brianlovin** [comment]
BOOM. @mxstbr https://github.com/withspectrum/spectrum/pull/4751/commits/7a31670f19013630f30ce8ebb3989e4af44b2c2c makes the composer behave the same as the thread slider, using the router and modal state. #feelsgoodman - the main benefit is that previously we were loading a hidden composer component on *every* page, now it only renders when it actually gets opened. This will also solve the conflicting event listeners for `esc` - going to push up a commit shortly with tests for the composer and thread slider esc handlers.

**Comment by mxstbr** [comment]
Dude that's amazing :tada: :tada: Can't wait to look at this tomorrow!

**Comment by brianlovin** [comment]
Ended up doing a lot more work here than anticipated @mxstbr - specifically added a ton of tests for the routing behavior (including tests for the thread slider routing).

I also:
- emphasize the community/channel dropdowns if nothing is selected yet
- auto select the first channel in the dropdown if the community only has one channel (eg general)

Also found a bunch of empty feed states that needed cleaning up and proper composer routing. I think I tracked down everything!

**Comment by brianlovin** [comment]
Just realized we should probably also auto-select the first community in the dropdown if the user is only a member of one community. Should be a fast follow in the morning, I'm done for tonight, unless you wanna tackle it based on https://github.com/withspectrum/spectrum/pull/4751/commits/6525f3f96ae561d5823673286c8acbf848d1f5c6 and https://github.com/withspectrum/spectrum/pull/4751/commits/a8e6cacb8ecdd9b3788ba62375947695710256d6 (not sure if we'll need a new seed user that is a member of only one community, or if we already have one that would work)

**Comment by brianlovin** [comment]
Okay, last commit now supports:

- if a user is only a member of a single community, we select that community by default
- if that community only has one channel, we will select that single channel by default

Smarter defaults, yay!

**Comment by brianlovin** [comment]
These e2e tests will be the death of me...

**Comment by brianlovin** [comment]
I have to skip these...every test passes locally no matter how many times I run it. I cannot repro at all. I've been banging my head on these for 2 days now.

**Comment by brianlovin** [comment]
screenshot 2019-03-01 15 29 29

I can't even with these tests...

**Comment by brianlovin** [comment]
screenshot 2019-03-01 15 34 47

Crying rn. Admin merging, gonna see if I can get all the other PRs to land.

**Pull request merged**